go/login/login.go

```diff
@@ -0,0 +1,388 @@
+}
+
+// GetTag returns the tag corresponding to the Msg. The returned tag is calculated with usage of sendingKey.
+func (r *URLResponse) GetTag(len uint) []byte {
+	return r.d.Tag(r.Msg.Construct(false), 0)[:len]
+}
+
+// GetEncToken returns a base64-encoded response token.
+func (r *URLResponse) GetEncToken() string {
+	return base64.URLEncoding.EncodeToString(r.GetTag(glome.MaxTagSize)) // TODO: passing the tag len as param?
+}
+
+// Client side glome-login handler. Should be constructed under NewClient constructor.
+type Client struct {
+	ServiceKey   glome.PublicKey
+	UserKey      glome.PrivateKey
+	ServiceKeyID uint8
+	TagLen       uint
+	response     *URLResponse
+}
+
+// NewClient is a Client constructor. Sets Client.ServiceKey, Client.UserKey, Client.ServiceKeyID, Client.TagLen
+// to the corresponding values and Client.response to nil.
+func NewClient(sk glome.PublicKey, uk glome.PrivateKey, sID uint8, tagLen uint) *Client {
+	return &Client{sk, uk, sID, tagLen, nil}
+}
+
+// Construct returns a request to the server according to the format: /v<V>/<glome-handshake>[/<message>]/.
+func (c *Client) Construct(V byte, hostIDType string, hostID string, action string) (string, error) {
+	r, err := NewResponse(c.ServiceKeyID, c.ServiceKey, c.UserKey, V, hostIDType, hostID, action, c.TagLen)
+	if err != nil {
+		return "", err
+	}
+	c.response = r
+
+	var handshake = c.constructHandshake()
+	var msg = c.response.Msg.Construct(true)
+	var u = fmt.Sprintf("/v%d/%s/", c.response.V, handshake)
+	if len(msg) > 0 {
+		u += fmt.Sprintf("%s/", msg)
+	}
+	return u, nil
+}
+
+// constructHandshake returns base64-url encoded handshake. The handshake is constructed following the format:
+//		glome-handshake := base64url(
+//    		<prefix-type>
+//    		<prefix7>
+//    		<eph-key>
+//    		[<prefixN>]
+//  	).
+func (c *Client) constructHandshake() string {
+	var handshake []byte
+	h := c.response.HandshakeInfo
+
+	handshake = append(handshake, h.Prefix)
+	handshake = append(handshake, h.UserKey[:]...)
+	handshake = append(handshake, h.MsgTagPrefix[:]...)
+	return base64.URLEncoding.EncodeToString(handshake[:])
+}
+
+// ValidateAuthCode checks if the received tag corresponding to the base64-url encoded message constructed from the Message.
+func (c *Client) ValidateAuthCode(tag string) (bool, error) {
+	dTag, err := base64.URLEncoding.DecodeString(completeBase64S(tag))
+	if err != nil {
+		return false, err
+	}
+
+	if c.response == nil {
+		return false, ErrResponseNotInitialized
+	}
+	return c.response.ValidateAuthCode(dTag), nil
+}
+
+// completeBase64S completes the base64 string with padding if it was truncated and couldn't be correctly decoded.
+func completeBase64S(s string) string {
+	n := len(s)
+	switch n % 4 {
+	case 0:
+		return s
+	case 1:
+		return s[:n-1]
+	case 2:
+		return s + "=="
+	case 3:
+		return s + "="
+	default:
+		panic("math fail")
+	}
+}
+
+func (c *Client) getResponse() *URLResponse {
```

Comment:
https://golang.org/doc/effective_go.html#Getters

Reply:
Also apply to getTag

Reply:
Fixed.